code/client/src/pages/Restore.jsx

```diff
@@ -128,4 +129,134 @@ const Restore = () => {
     </>
   )
 }
-export default Restore
+
+const RestoreV2 = () => {
+  const history = useHistory()
+  const [section, setSection] = useState(Sections.Choose)
+
+  const [walletInfo, setWalletInfo] = useState()
+
+  const [progress, setProgress] = useState(0)
+  const [progressStage, setProgressStage] = useState(0)
+  const [innerTrees, setInnerTrees] = useState()
+  const [innerCores, setInnerCores] = useState()
+  const [expert, setExpert] = useState()
+  const [name, setName] = useState()
+  const [newLocalParams, setNewLocalParams] = useState()
+  const [address, setAddress] = useState()
+
+  const onSynced = async ({ name, address: retrievalAddress, innerTrees, expert }) => {
+    try {
+      const { wallet } = await retrieveWalletInfoFromAddress(retrievalAddress)
```

Comment:
if all the set ups are the same, like the only change is the dom structure, I would do the upgrade support from the Restore component. Of course you will need to extract the components from Restore to multiple components first (like per section), but that should make it more readable and modular as well ;) we should try to avoid very large components in general, they tend to become really hard to maintain pretty soon, especially when no tests ;)